From 33b9c6aeed551e70f072d597c65bfff66fc8c6e3 Mon Sep 17 00:00:00 2001
From: lei zhilin <leizhilin@eswincomputing.com>
Date: Wed, 5 Mar 2025 17:34:40 +0800
Subject: [PATCH 438/652] <fix>(2d):2d isr interrupt call muetx err

Changelogs:

1.Problem cause:
    when call drv_release by user, gckKERNEL_AttachProcessEx call gcvBROADCAST_LAST_PROCESS,
    send poweroff cmd, power off clock, but interrupt come now, call gcvPOWER_ON,
    use sleep function cause this problem

2.Commit that introduced the problem:
    http://10.10.130.10:8080/c/win2030/linux-6.6/+/16385
    http://10.10.130.10:8080/c/win2030/linux-6.6/+/16241

3.Solutions:
  1) revert commit, delete power on call in interrupt;
  2) revert commit, no need judge core 0 or 1 in gckKERNEL_AttachProcessEx;
  3) use gcdENABLE_PER_DEVICE_PM to not seed power off, because powerStateTimer will triger to do powerff delay

Signed-off-by: lei zhilin <leizhilin@eswincomputing.com>
---
 .../eswin/hae/hal/kernel/gc_hal_kernel.c      |  6 -----
 .../os/linux/kernel/gc_hal_kernel_device.c    |  8 +-----
 .../hal/os/linux/kernel/gc_hal_kernel_os.c    | 26 +++++++++++++++++++
 3 files changed, 27 insertions(+), 13 deletions(-)

diff --git a/drivers/staging/media/eswin/hae/hal/kernel/gc_hal_kernel.c b/drivers/staging/media/eswin/hae/hal/kernel/gc_hal_kernel.c
index fb921fbd970e..c5b7110e346d 100644
--- a/drivers/staging/media/eswin/hae/hal/kernel/gc_hal_kernel.c
+++ b/drivers/staging/media/eswin/hae/hal/kernel/gc_hal_kernel.c
@@ -3957,12 +3957,6 @@ gckKERNEL_AttachProcessEx(gckKERNEL Kernel, gctBOOL Attach, gctUINT32 PID)
                 /* Last client detached, switch to SUSPEND power state. */
                 gcmkONERROR(gckOS_Broadcast(Kernel->os, Kernel->hardware,
                                             gcvBROADCAST_LAST_PROCESS));
-                if (Kernel->core == gcvCORE_2D1 && Kernel->hardware->options.powerManagement) {
-                    /*notify 2d core0 to power off*/
-                    gckKERNEL ker = gcvNULL;
-                    gckOS_QueryKernel(Kernel, gcvCORE_2D, &ker);
-                    gcmkONERROR(gckOS_Broadcast(ker->os, ker->hardware, gcvBROADCAST_LAST_PROCESS));
-                }
 
                 if (Kernel->processPageTable) {
                     status = gckEVENT_Submit(Kernel->eventObj, &eventAttr);
diff --git a/drivers/staging/media/eswin/hae/hal/os/linux/kernel/gc_hal_kernel_device.c b/drivers/staging/media/eswin/hae/hal/os/linux/kernel/gc_hal_kernel_device.c
index b31682bb944e..9699a7b40d14 100644
--- a/drivers/staging/media/eswin/hae/hal/os/linux/kernel/gc_hal_kernel_device.c
+++ b/drivers/staging/media/eswin/hae/hal/os/linux/kernel/gc_hal_kernel_device.c
@@ -2267,14 +2267,8 @@ isrRoutine(int irq, void *ctxt)
         return IRQ_HANDLED;
     }
 
-    /* interrupt came when the clock disable, so set power on to handle interrupt */
     if(status == gcvSTATUS_GENERIC_IO){
-        gckHARDWARE_SetPowerState(kernel->hardware, gcvPOWER_ON_AUTO);
-        status = gckHARDWARE_Interrupt(kernel->hardware);
-        if (gcmIS_SUCCESS(status)) {
-            up(kernel->sema);
-            return IRQ_HANDLED;
-        }
+        pr_info("[%s] hae isr meet clock close...\n", __func__);
     }
 
     return IRQ_NONE;
diff --git a/drivers/staging/media/eswin/hae/hal/os/linux/kernel/gc_hal_kernel_os.c b/drivers/staging/media/eswin/hae/hal/os/linux/kernel/gc_hal_kernel_os.c
index 68bcf61c95f0..e661e24e2fb5 100644
--- a/drivers/staging/media/eswin/hae/hal/os/linux/kernel/gc_hal_kernel_os.c
+++ b/drivers/staging/media/eswin/hae/hal/os/linux/kernel/gc_hal_kernel_os.c
@@ -1,3 +1,26 @@
+// SPDX-License-Identifier: GPL-2.0
+/******************************************************************************
+ *
+ * ESWIN hae driver
+ *
+ * Copyright 2024, Beijing ESWIN Computing Technology Co., Ltd.. All rights reserved.
+ *
+ * This program is free software: you can redistribute it and/or modify
+ * it under the terms of the GNU General Public License as published by
+ * the Free Software Foundation, version 2.
+ *
+ * This program is distributed in the hope that it will be useful,
+ * but WITHOUT ANY WARRANTY; without even the implied warranty of
+ * MERCHANTABILITY or FITNESS FOR A PARTICULAR PURPOSE.  See the
+ * GNU General Public License for more details.
+ *
+ * You should have received a copy of the GNU General Public License
+ * along with this program.  If not, see <https://www.gnu.org/licenses/>.
+ *
+ * Authors: zhilin lei <leizhilin@eswincomputing.com>
+ *
+ ***************************************************************************/
+
 /****************************************************************************
 *
 *    The MIT License (MIT)
@@ -4292,7 +4315,10 @@ gckOS_Broadcast(gckOS Os, gckHARDWARE Hardware, gceBROADCAST Reason)
         gcmkTRACE_ZONE(gcvLEVEL_INFO, gcvZONE_OS, "Last process has detached");
 
         /* Put GPU OFF. */
+        /* With PM, Clock will power off by powerStateTimer*/
+#if !gcdENABLE_PER_DEVICE_PM
         gcmkONERROR(gckHARDWARE_SetPowerState(Hardware, gcvPOWER_OFF_BROADCAST));
+#endif
         break;
 
     case gcvBROADCAST_GPU_IDLE:
-- 
2.53.0

